### PR TITLE
Refactor CA trust store check method

### LINKED ISSUE #17576: [Hosted Che] update to 7.17.0 is failing due to changes in Openshift4TrustedCAProvisioner
### Describe the bug
Provision custom CA certificates into every container of the workspace pod have been implemented as part of:
- https://github.com/eclipse/che/issues/17407
- https://github.com/eclipse/che/pull/17460

Currently, the changes prevent Hosted Che update to 7.17.0 upstream version. Basically, with the current implementation, it is not possible to opt-out from the usage of the `Openshift4TrustedCAProvisioner` (cm check should not happen for Hosted Che case since SA has no permissions to do it). In general, there should be a possibility to disable custom CA certificates into every container of the workspace since this case is addressing a particular case with Eclipse Che deployed behind the proxy.)

### Che version
7.17.0

### Expected behavior
Successful Hosted Che deployment against the latest 7.17.0 upstream
for Hosted Che case Openshift4TrustedCAProvisioner should not be called - https://github.com/eclipse/che/pull/17460/files#diff-1608a763bf13a812bb359246b8a62d96R52

### Runtime
   - [x] Openshift (include output of `oc version`)

### Environment
   - [x] OSD v4 che-dev cluster

### Eclipse Che Logs
```
2020-08-04 15:27:47,167[ost-startStop-1]  [ERROR] [o.a.c.c.C.[.[localhost].[/api] 175]  - Exception sending context initialized event to listener instance of class [org.eclipse.che.inject.CheBootstrap]
com.google.inject.CreationException: Unable to create injector, see the following errors:
1) Error injecting constructor, io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.30.0.1/api/v1/namespaces/prcheck-1959/configmaps?labelSelector=config.openshift.io%2Finject-trusted-cabundle%3Dtrue. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. configmaps is forbidden: User "system:serviceaccount:prcheck-1959:rhche" cannot list resource "configmaps" in API group "" in the namespace "prcheck-1959".
  at org.eclipse.che.workspace.infrastructure.openshift.provision.Openshift4TrustedCAProvisioner.<init>(Openshift4TrustedCAProvisioner.java:58)
  at org.eclipse.che.workspace.infrastructure.openshift.provision.Openshift4TrustedCAProvisioner.class(Openshift4TrustedCAProvisioner.java:58)
  while locating org.eclipse.che.workspace.infrastructure.openshift.provision.Openshift4TrustedCAProvisioner
1 error
	at com.google.inject.internal.Errors.throwCreationExceptionIfErrorsExist(Errors.java:543)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:186)
	at com.google.inject.internal.InternalInjectorCreator.build(InternalInjectorCreator.java:109)
	at com.google.inject.Guice.createInjector(Guice.java:87)
	at org.everrest.guice.servlet.EverrestGuiceContextListener.getInjector(EverrestGuiceContextListener.java:141)
	at com.google.inject.servlet.GuiceServletContextListener.contextInitialized(GuiceServletContextListener.java:45)
	at org.everrest.guice.servlet.EverrestGuiceContextListener.contextInitialized(EverrestGuiceContextListener.java:86)
	at org.apache.catalina.core.StandardContext.listenerStart(StandardContext.java:4689)
	at org.apache.catalina.core.StandardContext.startInternal(StandardContext.java:5155)
	at org.apache.catalina.util.LifecycleBase.start(LifecycleBase.java:183)
	at org.apache.catalina.core.ContainerBase.addChildInternal(ContainerBase.java:743)
	at org.apache.catalina.core.ContainerBase.addChild(ContainerBase.java:719)
	at org.apache.catalina.core.StandardHost.addChild(StandardHost.java:705)
	at org.apache.catalina.startup.HostConfig.deployWAR(HostConfig.java:970)
	at org.apache.catalina.startup.HostConfig$DeployWar.run(HostConfig.java:1840)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:834)
Caused by: io.fabric8.kubernetes.client.KubernetesClientException: Failure executing: GET at: https://172.30.0.1/api/v1/namespaces/prcheck-1959/configmaps?labelSelector=config.openshift.io%2Finject-trusted-cabundle%3Dtrue. Message: Forbidden!Configured service account doesn't have access. Service account may have been revoked. configmaps is forbidden: User "system:serviceaccount:prcheck-1959:rhche" cannot list resource "configmaps" in API group "" in the namespace "prcheck-1959".
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.requestFailure(OperationSupport.java:568)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.assertResponseCode(OperationSupport.java:505)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:471)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:430)
	at io.fabric8.kubernetes.client.dsl.base.OperationSupport.handleResponse(OperationSupport.java:412)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.listRequestHelper(BaseOperation.java:151)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:621)
	at io.fabric8.kubernetes.client.dsl.base.BaseOperation.list(BaseOperation.java:70)
	at org.eclipse.che.workspace.infrastructure.openshift.provision.Openshift4TrustedCAProvisioner.<init>(Openshift4TrustedCAProvisioner.java:68)
	at org.eclipse.che.workspace.infrastructure.openshift.provision.Openshift4TrustedCAProvisioner$$FastClassByGuice$$afa88f6b.newInstance(<generated>)
	at com.google.inject.internal.DefaultConstructionProxyFactory$FastClassProxy.newInstance(DefaultConstructionProxyFactory.java:89)
	at com.google.inject.internal.ConstructorInjector.provision(ConstructorInjector.java:114)
	at com.google.inject.internal.ConstructorInjector.construct(ConstructorInjector.java:91)
	at com.google.inject.internal.ConstructorBindingImpl$Factory.get(ConstructorBindingImpl.java:306)
	at com.google.inject.internal.ProviderToInternalFactoryAdapter.get(ProviderToInternalFactoryAdapter.java:40)
	at com.google.inject.internal.SingletonScope$1.get(SingletonScope.java:168)
	at com.google.inject.internal.InternalFactoryToProviderAdapter.get(InternalFactoryToProviderAdapter.java:39)
	at com.google.inject.internal.InternalInjectorCreator.loadEagerSingletons(InternalInjectorCreator.java:211)
	at com.google.inject.internal.InternalInjectorCreator.injectDynamically(InternalInjectorCreator.java:182)
	... 18 common frames omitted
04-Aug-2020 15:27:47.169 SEVERE [localhost-startStop-1] org.apache.catalina.core.StandardContext.startInternal One or more listeners failed to start. Full details will be found in the appropriate container log file
04-Aug-2020 15:27:47.169 SEVERE [localhost-startStop-1] org.apache.catalina.core.StandardContext.startInternal Context [/api] startup failed due to previous errors
```
### Additional context
Hosted Che PR check https://github.com/redhat-developer/rh-che/pull/1950

### PR DESCRIPTION
**Pull request by mshaposhnik**
### What does this PR do?
Due to some kind of permissions errors, refactored method of checking if CA store is initialized or not.

### What issues does this PR fix or reference?

Fixes https://github.com/eclipse/che/issues/17576

#### Docs PR
N/A

### CODE REVIEW

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/Openshift4TrustedCAProvisioner.java [hunk 1]
```diff
@@ -50,24 +48,15 @@
 
   @Inject
   public Openshift4TrustedCAProvisioner(
+      @Nullable @Named("che.trusted_ca_bundles_configmap") String caBundleConfigMap,
```
Comment (sparkoo) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/Openshift4TrustedCAProvisioner.java:51]: from where does this value come from?

**Review by sparkoo — COMMENTED** [review]
I don't much understand what's happening here. The code before was checking some actual configmaps are present on k8s, which looks reasonable for `trustedStoreInitialized` variable. Maybe your code is doing the same, I just don't know what is the source of `@Nullable @Named("che.trusted_ca_bundles_configmap") String caBundleConfigMap`

**Comment by skabashnyuk** [comment]
CHE_TRUSTED__CA__BUNDLES__CONFIGMAP os not defined

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/Openshift4TrustedCAProvisioner.java [hunk 2]
```diff
@@ -11,26 +11,24 @@
...
 import org.eclipse.che.api.workspace.server.spi.InfrastructureException;
+import org.eclipse.che.commons.annotation.Nullable;
 import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment;
 import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodData;
 import org.eclipse.che.workspace.infrastructure.kubernetes.environment.KubernetesEnvironment.PodRole;
-import org.eclipse.che.workspace.infrastructure.openshift.OpenShiftClientFactory;
-import org.eclipse.che.workspace.infrastructure.openshift.environment.OpenShiftCheInstallationLocation;
 import org.eclipse.che.workspace.infrastructure.openshift.project.OpenShiftProject;
 
 /**
```
Comment (skabashnyuk) [inline infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/Openshift4TrustedCAProvisioner.java:34]: can you explain in javadoc how logic with trusted_ca_bundles_configmap works? 
Explain why https://github.com/eclipse/che/issues/17576#issuecomment-669130679 this should work.

File: infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/Openshift4TrustedCAProvisioner.java (hunk 1 above)
Comment (mshaposhnik) [reply infrastructures/openshift/src/main/java/org/eclipse/che/workspace/infrastructure/openshift/provision/Openshift4TrustedCAProvisioner.java:51]: missed to add che.props :), now added

**Comment by sparkoo** [comment]
so does it really mean if `che.trusted_ca_bundles_configmap` is set to whatever value, that trusted store is initialized? That's what code is saying, but it does not looks correct?
Who is responsible for creating that configmap? What if value is set but actual configmap is not there?

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 3]
```diff
@@ -49,6 +49,9 @@ che.workspace.http_proxy=
 che.workspace.https_proxy=
 che.workspace.no_proxy=
 
+# Configures CA bundle config map name for cluster-wide proxy mode
+che.trusted_ca_bundles_configmap=NULL
```
Comment (ibuziuk) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:58]: I am really missing the clarity in the docs here. Would be nice to describe what is the cm, why / when it should be set, how cluster-wide proxy mode will work etc.

Comment (mshaposhnik) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:58]: Added more explanation from Anatoly

Comment (sparkoo) [reply assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:58]: Can you please write a little note/warning that's clearly says this is advanced property and it's not expected to be manually set unless you know really well what you're doing.

**Review by sparkoo — APPROVED** [review]

**Comment by mshaposhnik** [comment]
[crw-ci-test]

**Review by ibuziuk — APPROVED** [review]
@mshaposhnik great! I guess it will not be part of 7.17.0 (could you please also cherry-pick it to 7.17.x)

File: assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties [hunk 4]
```diff
@@ -49,6 +49,13 @@ che.workspace.http_proxy=
 che.workspace.https_proxy=
 che.workspace.no_proxy=
 
+# When cluster wide proxy is configured, che-operator creates special configmap
+# and allows OpenShift Network operator to inject ca-bundle into it.
+# In addition, it adds the key CHE_TRUSTED__CA__BUNDLES__CONFIGMAP with name of this configmap
```
Comment (skabashnyuk) [inline assembly/assembly-wsmaster-war/src/main/webapp/WEB-INF/classes/che/che.properties:54]: it adds environment variable `CHE_TRUSTED__CA__BUNDLES__CONFIGMAP`

**Review by skabashnyuk — APPROVED** [review]

**Pull request merged**